akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala

```diff
@@ -511,6 +523,32 @@ private[persistence] trait Eventsourced extends Snapshotter with Stash with Stas
     }
   }
 
+  private def flushBatch() {
+    // When using only `persistAsync` and `defer` max throughput is increased by using
+    // batching, but when using `persist` we want to use one atomic WriteMessages
+    // for the emitted events.
+    // Flush previously collected events, if any, separately from the `persist` batch
+    if (pendingStashingPersistInvocations > 0 && journalBatch.nonEmpty)
+      flushJournalBatch()
+
+    eventBatch.reverse.foreach { p ⇒
+      addToBatch(p)
+      if (!writeInProgress || maxBatchSizeReached) flushJournalBatch()
+    }
+
+    eventBatch = Nil
+  }
+
+  private def addToBatch(p: PersistentEnvelope): Unit = p match {
+    case p: PersistentRepr ⇒
+      journalBatch :+= p.update(persistenceId = persistenceId, sequenceNr = nextSequenceNr(), sender = sender())
+    case r: PersistentEnvelope ⇒
+      journalBatch :+= r
+  }
+
+  private def maxBatchSizeReached: Boolean =
+    journalBatch.size >= maxMessageBatchSize
+
```

Comment:
yes, it's just moved.
It was inside a `State`, but now it's possible to add to a batch from different places (because `persist()(persist)`)